src/java.base/share/classes/java/lang/StringUTF16.java

```diff
@@ -1541,14 +1576,23 @@ static int getChars(int i, int index, byte[] buf) {
             q = i / 100;
             r = (q * 100) - i;
             i = q;
-            putChar(buf, --charPos, Integer.DigitOnes[r]);
-            putChar(buf, --charPos, Integer.DigitTens[r]);
+
+            charPos -= 2;
+            UNSAFE.putIntUnaligned(
+                    buf,
+                    Unsafe.ARRAY_BYTE_BASE_OFFSET + (charPos << 1),
+                    PACKED_DIGITS_UTF16[r]);
         }
 
         // We know there are at most two digits left at this point.
-        putChar(buf, --charPos, Integer.DigitOnes[-i]);
         if (i < -9) {
-            putChar(buf, --charPos, Integer.DigitTens[-i]);
+            charPos -= 2;
+            UNSAFE.putIntUnaligned(
+                    buf,
+                    Unsafe.ARRAY_BYTE_BASE_OFFSET + (charPos << 1),
+                    PACKED_DIGITS_UTF16[-i]);
+        } else {
+            putChar(buf, --charPos, '0' - i);
```

Comment:
Ditto add explicit array bounds check when using Unsafe, especially since the method is used outside of the source file.  Here and in the uses of Unsafe below.

Reply:
I added assert with reference to the implementation of StringUTF16#putChar, is this safe enough?